### PR TITLE
=htt #20338 correct byte range slicing

### LINKED ISSUE #20338: akka-http: InvalidContentLengthException with akka 2.4.4
Hi everybody,

I just tried to serve some binary data (a video) with akka-http and I got this exception:

```
app-backend 20:18:19.886 [actorSystem-akka.actor.default-dispatcher-3] DEBUG a.i.TcpListener - Successfully bound to /0:0:0:0:0:0:0:0:9100
app-backend 20:18:21.664 [actorSystem-akka.actor.default-dispatcher-2] DEBUG a.i.TcpListener - New connection accepted
app-backend 20:18:22.014 [actorSystem-akka.actor.default-dispatcher-4] DEBUG a.i.TcpIncomingConnection - Closing connection due to IO error java.io.IOException: Connection reset by peer
app-backend 20:18:22.016 [actorSystem-akka.actor.default-dispatcher-2] DEBUG a.i.TcpListener - New connection accepted
app-backend 20:18:22.025 [actorSystem-akka.actor.default-dispatcher-23] INFO  a.a.LocalActorRef - Message [akka.io.Tcp$Write] from Actor[akka://actorSystem/user/StreamSupervisor-0/$$b#-808484062] to Actor[akka://actorSystem/system/IO-TCP/selectors/$a/1#1730712293] was not delivered. [1] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
app-backend 20:18:22.421 [actorSystem-akka.actor.default-dispatcher-22] DEBUG a.s.i.i.InputStreamPublisher - No more bytes available to read (got `-1` from `read`)
app-backend 20:18:22.434 [actorSystem-akka.actor.default-dispatcher-2] ERROR a.a.ActorSystemImpl - Outgoing response stream error
app-backend akka.http.scaladsl.model.InvalidContentLengthException: HTTP message had declared Content-Length 25656027 but entity data stream amounts to 11323995 bytes less
app-backend 	at akka.http.scaladsl.model.InvalidContentLengthException$.apply(ErrorInfo.scala:50) ~[akka-http-core_2.11-2.4.4.jar:na]
app-backend 	at akka.http.impl.engine.rendering.RenderSupport$CheckContentLengthTransformer$$anon$1.onUpstreamFinish(RenderSupport.scala:97) ~[akka-http-core_2.11-2.4.4.jar:na]
app-backend 	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:625) ~[akka-stream_2.11-2.4.4.jar:na]
app-backend 	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:539) ~[akka-stream_2.11-2.4.4.jar:na]
app-backend 	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:469) ~[akka-stream_2.11-2.4.4.jar:na]
app-backend 	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:431) ~[akka-stream_2.11-2.4.4.jar:na]
app-backend 	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:593) ~[akka-stream_2.11-2.4.4.jar:na]
app-backend 	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:604) ~[akka-stream_2.11-2.4.4.jar:na]
app-backend 	at akka.actor.Actor$class.aroundReceive(Actor.scala:482) ~[akka-actor_2.11-2.4.4.jar:na]
app-backend 	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:519) ~[akka-stream_2.11-2.4.4.jar:na]
app-backend 	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526) [akka-actor_2.11-2.4.4.jar:na]
app-backend 	at akka.actor.ActorCell.invoke(ActorCell.scala:495) [akka-actor_2.11-2.4.4.jar:na]
app-backend 	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257) [akka-actor_2.11-2.4.4.jar:na]
app-backend 	at akka.dispatch.Mailbox.run(Mailbox.scala:224) [akka-actor_2.11-2.4.4.jar:na]
app-backend 	at akka.dispatch.Mailbox.exec(Mailbox.scala:234) [akka-actor_2.11-2.4.4.jar:na]
app-backend 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [na:na]
app-backend 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [na:na]
app-backend 	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [na:na]
app-backend 	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [na:na]
app-backend 20:18:22.435 [actorSystem-akka.actor.default-dispatcher-2] DEBUG a.a.RepointableActorRef - Aborting tcp connection because of upstream failure: HTTP message had declared Content-Length 25656027 but entity data stream amounts to 11323995 bytes less
app-backend akka.http.scaladsl.model.InvalidContentLengthException$.apply(ErrorInfo.scala:50)
app-backend akka.http.impl.engine.rendering.RenderSupport$CheckContentLengthTransformer$$anon$1.onUpstreamFinish(RenderSupport.scala:97)
app-backend akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:625)
app-backend akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:539)
app-backend akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:469)
app-backend akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:431)
app-backend akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:593)
app-backend akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:604)
app-backend akka.actor.Actor$class.aroundReceive(Actor.scala:482)
app-backend akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:519)
app-backend akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
app-backend akka.actor.ActorCell.invoke(ActorCell.scala:495)
app-backend akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
app-backend akka.dispatch.Mailbox.run(Mailbox.scala:224)
app-backend akka.dispatch.Mailbox.exec(Mailbox.scala:234)
app-backend scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
app-backend scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
app-backend scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
app-backend scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I used this Route definition to produce the error:
```scala
trait Routes {

  val route =
      get {
        path("media") {
          getFromResource("VID_20141108_124530.mp4")
        } ~
          pathEndOrSingleSlash {
            complete(OK, "Hello World")
          }
      }
}
```

I got the exception with akka 2.4.4 (Scala 2.11.8) on OS X 10.11.4. I also tried with akka 2.4.3, which worked.

If you need more information just let me know.

Matthias

### PR DESCRIPTION
**Pull request by aruediger**
fixes #20338

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala [hunk 1]
```diff
@@ -83,17 +83,15 @@ private[http] object StreamUtils {
...
           val element = grab(in)
-          if (toSkip > 0 && element.length < toSkip) {
-            // keep skipping
-            toSkip -= element.length
+          if (toSkip >= element.length) {
             pull(in)
           } else {
-            // toSkip <= element.length <= Int.MaxValue
-            val data = element.drop(toSkip.toInt).take(math.min(remaining, Int.MaxValue).toInt)
+            val data = element.drop(toSkip.toInt).take(remaining.toInt)
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: What if remaining > 2^31?

Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: Oops. Looks like I oversimplified here. Thx. Will fix.

**Comment by 2beaucoup** [comment]
updated

**Comment by jrudolph** [comment]
:+1:

**Comment by ktoso** [comment]
LGTM, thanks @2beaucoup !

**Comment by antonkulaga** [comment]
Great, how soon are you going to release next minor version of akka?

**Comment by ktoso** [comment]
Once new JavaDSL is done for Akka HTTP, so much quicker if someone would like to help... :-)
Needs help cleaning up akka-docs for javadsl for example (this is the main blocker): https://github.com/ktoso/akka/tree/jypma-java-http-server-dsl
Help would be very very welcome.

**Comment by ktoso** [comment]
More info on what's blocking the release in: https://github.com/akka/akka-meta/issues/17
Effectively 2.4.5 means "once javadsl is done", which is a large piece of boring work

**Comment by antonkulaga** [comment]
What about publishing a snapshot (I do not use JavaDSL so seems like it should not break anything) so I will be able to add it to my deps and check if it fixes this annoying problem?

**Comment by ktoso** [comment]
I can't now, but it's rather simple for yourself to try:

git clone https://github.com/akka/akka.git
cd akka
sbt publishLocal 

:-)

-- 
Konrad `ktoso` Malawski
Akka @ Lightbend

On 28 April 2016 at 19:52:03, Anton Kulaga (notifications@github.com) wrote:

What about publishing a snapshot (I do not use JavaDSL so seems like it should not break anything) so I will be able to add it to my deps and check if it fixes this annoying problem?

**Comment by patriknw** [comment]
Nightly snapshots are published automatically. Search docs for details

**Pull request merged**